File: 백준/Gold/21276. 계보 복원가 호석/계보 복원가 호석.java
import java.io.*;
import java.util.*;

public class Main {

    static int n;
    static int m;
    static List<String> names = new ArrayList<>();
    static TreeSet<String> ancestors = new TreeSet<>();
    static int deg[] = new int[1002];
    static ArrayList<ArrayList<Integer>> pgraph = new ArrayList<>();
    static ArrayList<ArrayList<Integer>> cgraph = new ArrayList<>();
    static Map<String, TreeSet<String>> childs = new TreeMap<>();

    public static int findIndexByName(String name) {
        for (int i = 0; i < n; i++) {
            if (names.get(i).equals(name)) return i;
        }
        return -1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();
        n = Integer.parseInt(br.readLine());
        for (int i = 0; i < n; i++) {
            pgraph.add(new ArrayList<>());
            cgraph.add(new ArrayList<>());
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) names.add(st.nextToken());
        m = Integer.parseInt(br.readLine());
        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = findIndexByName(st.nextToken());
            int v = findIndexByName(st.nextToken());
            pgraph.get(u).add(v);
            cgraph.get(v).add(u);
            deg[u]++;
        }
        Queue<Integer> queue = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            if (deg[i] == 0) queue.offer(i);
            if (pgraph.get(i).size() == 0) ancestors.add(names.get(i));
        }
        sb.append(ancestors.size()).append('\n');
        for (String name : ancestors) {
            sb.append(name).append(' ');
        }
        sb.append('\n');

        for (String name : names) {
            childs.put(name, new TreeSet<>());
        }

        while (!queue.isEmpty()) {
            int cur = queue.poll();

            for (int nxt : cgraph.get(cur)) {
                deg[nxt]--;
                if (deg[nxt] == 0) {
                    childs.get(names.get(cur)).add(names.get(nxt));
                    queue.offer(nxt);
                }
            }
        }

        for (String name : childs.keySet()) {
            TreeSet<String> list = childs.get(name);
            sb.append(name).append(' ').append(list.size()).append(' ');
            for (String child : list) sb.append(child).append(' ');
            sb.append('\n');
        }
        System.out.print(sb);
    }
}

File: 백준/Gold/21276. 계보 복원가 호석/계보 복원가 호석.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val names = StringTokenizer(readLine()).let { st -> List(n) { st.nextToken() } }.sorted()
    val nameMap = mutableMapOf<String, Int>()
    names.withIndex().forEach { (idx, v) -> nameMap[v] = idx }

    val adj = Array(n + 1) { mutableListOf<Int>() }
    val indeg = IntArray(n + 1)

    repeat(readLine().toInt()) {
        val st = StringTokenizer(readLine())
        val u = nameMap[st.nextToken()] ?: -1
        val v = nameMap[st.nextToken()] ?: -1
        adj[v] += u
        indeg[u]++
    }

    val rootIds = (0 until n).filter { indeg[it] == 0 }
    println(rootIds.size)
    println(rootIds.map { names[it] }.joinToString(" "))

    val childs = Array(n + 1) { TreeSet<Int>() }
    val queue = ArrayDeque<Int>()
    rootIds.forEach { queue.offer(it) }

    while (queue.isNotEmpty()) {
        var cur = queue.poll()

        for (nxt in adj[cur]) {
            if (--indeg[nxt] == 0) {
                childs[cur].add(nxt)
                queue.offer(nxt)
            }
        }
    }

    val sb = StringBuilder()
    for ((id, name) in names.withIndex()) {
        sb.append(name).append(' ').append(childs[id].size).append(' ')
        childs[id].forEach { sb.append(names[it]).append(' ') }
        sb.append('\n')
    }
    print(sb)
}